pkg/vault/vault.go

```diff
@@ -87,58 +89,99 @@ func (v *vaultApi) getClient() (*api.Client, error) {
 	return api.NewClient(config)
 }
 
-func (v *vaultApi) tokenTTL() (int64, error) {
-	resp, err := v.Request(http.MethodGet, SelfTokenURL, nil)
-	if err != nil || resp == nil {
-		return 0, fmt.Errorf("couldn't obtain self token information: %v", err)
+// A token is considered invalid if we receive 400 status codes
+func (v *vaultApi) tokenTTL() (ttl int64, invalid bool, err error) {
+	s, resp, err := v.Request(http.MethodGet, SelfTokenURL, nil)
+	if resp != nil && (resp.StatusCode == 400 || resp.StatusCode == 403) {
+		return 0, true, err
 	}
-	ttlNumber, ok := resp.Data["ttl"].(json.Number)
+	if err != nil {
+		return 0, false, fmt.Errorf("couldn't obtain self token information: %v", err)
+	}
+	ttlNumber, ok := s.Data["ttl"].(json.Number)
 	if !ok {
-		return 0, fmt.Errorf("couldn't obtain token TTL")
+		return 0, false, fmt.Errorf("couldn't obtain token TTL")
+	}
+	ttl, err = ttlNumber.Int64()
+	return ttl, false, err
+}
+
+// A token is not considered renewable if we receive a 400 error
+// for any other case we consider that it can still be renewed
+// and we are having problems connecting to the server.
+func (v *vaultApi) renewToken() (renewable bool, err error) {
+	s, resp, err := v.Request(http.MethodPost, SelfTokenRenewURL, nil)
+	if resp != nil && (resp.StatusCode == 400 || resp.StatusCode == 403) {
+		return false, err
+	}
+	if s != nil && s.Auth != nil {
+		return s.Auth.Renewable, err
+	}
+	if s != nil {
+		return s.Renewable, err
 	}
-	return ttlNumber.Int64()
+	return true, err
 }
 
 func (v *vaultApi) autoRenewToken() {
-	var ttl int64
-	var err error
+	const (
+		stateUpdateTTL = iota
+		stateRenew
+	)
 
-	retry := func(attempts int, t time.Duration, f func() error) {
-		for attempt := 0; attempt < attempts; attempt++ {
-			err = f()
-			if err == nil {
+	var state = stateUpdateTTL
```

Comment:
I think that for consistency we should probably agree on how we want to initialize variables inside functions. We do it in other places like: 

`state := stateUpdateTTL`

Reply:
Umm, in general I'd agree with that, but here I'm not so sure. We have two variables, and the other one doesn't need to be initialized.

What option would you prefer? All of them are valid, first one is the current one, second one is your proposal:
```
    var state = stateUpdateTTL
    var next time.Duration
```
```
    state := stateUpdateTTL
    var next time.Duration
```
```
    var state uint = stateUpdateTTL
    var next time.Duration
```
```
    state := stateUpdateTTL
    next := 0 * time.Second
```
```
    var ( 
        state = stateUpdateTTL
        next  time.Duration
    ) 
```

This would be also valid, but I'd prefer to initialize the state explicitely in this case:
```
    var state uint
    var next time.Duration
```

Reply:
Changed to your proposal :)